pyth_observer/check/stall_detection.py

```diff
@@ -0,0 +1,149 @@
+            min_noise_updates: Minimum number of updates needed for noise detection
+              (doesn't apply to exact stall detection)
+        """
+        self.stall_time_limit = stall_time_limit
+        self.noise_threshold = noise_threshold
+        self.min_noise_samples = min_noise_samples
+
+    def analyze_updates(self, updates: List[PriceUpdate]) -> StallDetectionResult:
+        """
+        Assumes that the cache has been recently updated since it takes the latest
+        cached timestamp as the current time.
+
+        Args:
+            updates: List of price updates to analyze
+
+        Returns:
+            StallDetectionResult with detection details
+        """
+        # Need at least 2 samples
+        if not updates or len(updates) < 2:
+            return StallDetectionResult.no_stall()
+
+        ## Check for exact stall
+
+        # The latest 2 updates are sufficient to detect an exact stall
+        latest_updates = updates[-2:]
+        duration = latest_updates[1].timestamp - latest_updates[0].timestamp
+        if duration <= self.stall_time_limit:
+            return StallDetectionResult.no_stall()
+        elif latest_updates[1].price == latest_updates[0].price:
+            return StallDetectionResult(
+                is_stalled=True,
+                stall_type="exact",
+                base_price=latest_updates[1].price,
+                noise_magnitude=0.0,
+                duration=duration,
+                confidence=1.0,
+            )
+
+        ## Check for stalled price with artificial noise added in
+
+        # Calculate relative deviations from base price
+        prices = np.array([u.price for u in updates])
+        base_price = np.median(prices)
+
+        if base_price == 0:
+            # Avoid division by zero
+            return StallDetectionResult.no_stall()
+
+        relative_deviations = np.abs(prices - base_price) / abs(base_price)
+        max_relative_deviation = np.max(relative_deviations)
+
+        # Check for artificial noise (variations below threshold)
+        if len(updates) < self.min_noise_samples:
+            # We need multiple samples to detect noise, pass until we have enough
+            return StallDetectionResult.no_stall()
+
+        if max_relative_deviation <= self.noise_threshold:
+            confidence = 1.0 - (max_relative_deviation / self.noise_threshold)
+            return StallDetectionResult(
+                is_stalled=True,
+                stall_type="noisy",
+                base_price=base_price,
+                noise_magnitude=max_relative_deviation * base_price,
+                duration=duration,
+                confidence=confidence,
+            )
+
+        return StallDetectionResult(
+            is_stalled=False,
+            stall_type=None,
+            base_price=base_price,
+            noise_magnitude=max_relative_deviation * base_price,
+            duration=duration,
+            confidence=0.0,
+        )
```

Comment:
```suggestion
        if base_price == 0:
            # Avoid division by zero
            return StallDetectionResult.no_stall()

        relative_deviations = np.abs(prices - base_price)
        max_relative_deviation = np.max(relative_deviations)

        # Check for artificial noise (variations below threshold)
        if len(updates) < self.min_noise_samples:
            # We need multiple samples to detect noise, pass until we have enough
            return StallDetectionResult.no_stall()

        if max_relative_deviation <= self.noise_threshold * base_price:
            confidence = 1.0 - (max_relative_deviation / self.noise_threshold)
            return StallDetectionResult(
                is_stalled=True,
                stall_type="noisy",
                base_price=base_price,
                noise_magnitude=max_relative_deviation,
                duration=duration,
                confidence=confidence,
            )

        return StallDetectionResult(
            is_stalled=False,
            stall_type=None,
            base_price=base_price,
            noise_magnitude=max_relative_deviation,
            duration=duration,
            confidence=0.0,
        )
```

Reply:
can you explain your suggestion pls? from what i gather you want to move from relative deviations (obtained by dividing by `base_price`) to discrete deviations. 

the reason I have `noise_threshold` and `max_relative_deviation` as "percent delta" or relative values is so that it works across different price scales -- it should work the same for both $0.1 prices and $100+.

Reply:
This will just avoid division so you don't have to worry about price being zero. I'll leave the judgement to you to see if it's better or not :D

Reply:
ahh i see - its a bit more intuitive for me with the division so will leave it :)